File: flutter/lib/remote.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';
import 'package:http/http.dart' as http;

String? _cactusToken;

void setCactusToken(String? token) {
  _cactusToken = token;
}

Future<List<double>> getVertexAIEmbedding(String text) async {
  if (_cactusToken == null) {
    throw Exception('CactusToken not set. Please call CactusLM.init with cactusToken parameter.');
  }

  const String projectId = 'cactus-v1-452518';
  const String location = 'us-central1';
  const String modelId = 'text-embedding-005';
  
  final String endpoint =
      'https://$location-aiplatform.googleapis.com/v1/projects/$projectId/locations/$location/publishers/google/models/$modelId:predict';

  final Map<String, String> headers = {
    'Authorization': 'Bearer $_cactusToken',
    'Content-Type': 'application/json',
  };

  final Map<String, dynamic> requestBody = {
    'instances': [
      {
        'content': text,
        'task_type': 'RETRIEVAL_DOCUMENT'
      }
    ]
  };

  try {
    final response = await http.post(
      Uri.parse(endpoint),
      headers: headers,
      body: json.encode(requestBody),
    );

    if (response.statusCode == 200) {
      final Map<String, dynamic> responseBody = json.decode(response.body);
      
      if (responseBody.containsKey('error')) {
        throw Exception('API Error: ${responseBody['error']['message']}');
      }
      
      final predictions = responseBody['predictions'] as List?;
      if (predictions == null || predictions.isEmpty) {
        throw Exception('No predictions in response');
      }
      
      final embeddings = predictions[0]['embeddings'];
      final values = embeddings['values'] as List;
      
      return values.map<double>((v) => v.toDouble()).toList();
    } else {
      throw Exception(
          'Failed to call Vertex AI. Status code: ${response.statusCode}\nBody: ${response.body}');
    }
  } catch (e) {
    throw Exception('An error occurred: $e');
  }
}

Future<String> getVertexAICompletion(
  String textPrompt, {
  Uint8List? imageData,
  String? imagePath,
  String? mimeType,
}) async {
  if (_cactusToken == null) {
    throw Exception('CactusToken not set. Please call CactusVLM.init with cactusToken parameter.');
  }

  const String projectId = 'cactus-v1-452518';
  const String location = 'global';
  const String modelId = 'gemini-2.5-flash-lite-preview-06-17';
  
  final String endpoint =
      'https://aiplatform.googleapis.com/v1/projects/$projectId/locations/$location/publishers/google/models/$modelId:generateContent';

  final Map<String, String> headers = {
    'Authorization': 'Bearer $_cactusToken',
    'Content-Type': 'application/json',
  };

  List<Map<String, dynamic>> parts = [
    {'text': textPrompt}
  ];

  if (imageData != null || imagePath != null) {
    Uint8List? finalImageData = imageData;
    String? finalMimeType = mimeType;

    if (imagePath != null && finalImageData == null) {
      final file = File(imagePath);
      if (await file.exists()) {
        finalImageData = await file.readAsBytes();
        
        if (finalMimeType == null) {
          final extension = imagePath.toLowerCase().split('.').last;
          switch (extension) {
            case 'jpg':
            case 'jpeg':
              finalMimeType = 'image/jpeg';
              break;
            case 'png':
              finalMimeType = 'image/png';
              break;
            case 'gif':
              finalMimeType = 'image/gif';
              break;
            case 'webp':
              finalMimeType = 'image/webp';
              break;
            default:
              finalMimeType = 'image/jpeg';
          }
        }
      } else {
        throw Exception('Image file not found: $imagePath');
      }
    }

    if (finalImageData != null) {
      final base64Image = base64Encode(finalImageData);
      parts.add({
        'inline_data': {
          'mime_type': finalMimeType ?? 'image/jpeg',
          'data': base64Image,
        }
      });
    }
  }

  final Map<String, dynamic> requestBody = {
    'contents': {
      'role': 'user',
      'parts': parts,
    }
  };

  try {
    final response = await http.post(
      Uri.parse(endpoint),
      headers: headers,
      body: json.encode(requestBody),
    );

    if (response.statusCode == 200) {
      final dynamic responseBody = json.decode(response.body);
      
      if (responseBody is Map && responseBody.containsKey('error')) {
        throw Exception('API Error: ${responseBody['error']['message']}');
      }
      
      dynamic candidates;
      if (responseBody is List) {
        candidates = responseBody;
      } else if (responseBody is Map && responseBody.containsKey('candidates')) {
        candidates = responseBody['candidates'];
      } else {
        throw Exception('Unexpected response format: ${response.body}');
      }
      
      if (candidates == null || candidates.isEmpty) {
        throw Exception('No candidates in response');
      }
      
      final content = candidates[0]['content'];
      final parts = content['parts'] as List;
      if (parts.isEmpty) {
        throw Exception('No parts in response');
      }
      
      final String modelOutput = parts[0]['text'] ?? '';
      return modelOutput;
    } else {
      throw Exception(
          'Failed to call Vertex AI. Status code: ${response.statusCode}\nBody: ${response.body}');
    }
  } catch (e) {
    throw Exception('An error occurred: $e');
  }
}

Future<String> getTextCompletion(String textPrompt) async {
  return getVertexAICompletion(textPrompt);
}

Future<String> getVisionCompletion(String textPrompt, String imagePath) async {
  return getVertexAICompletion(textPrompt, imagePath: imagePath);
}

Future<String> getVisionCompletionFromData(String textPrompt, Uint8List imageData, {String? mimeType}) async {
  return getVertexAICompletion(textPrompt, imageData: imageData, mimeType: mimeType);
}

File: react/src/remote.ts
let _cactusToken: string | null = null;

export function setCactusToken(token: string | null): void {
  _cactusToken = token;
}

export async function getVertexAIEmbedding(text: string): Promise<number[]> {
  if (_cactusToken === null) {
    throw new Error('CactusToken not set. Please call CactusLM.init with cactusToken parameter.');
  }

  const projectId = 'cactus-v1-452518';
  const location = 'us-central1';
  const modelId = 'text-embedding-005';
  
  const endpoint = `https://${location}-aiplatform.googleapis.com/v1/projects/${projectId}/locations/${location}/publishers/google/models/${modelId}:predict`;

  const headers = {
    'Authorization': `Bearer ${_cactusToken}`,
    'Content-Type': 'application/json',
  };

  const requestBody = {
    instances: [{ content: text }]
  };

  const response = await fetch(endpoint, {
    method: 'POST',
    headers,
    body: JSON.stringify(requestBody),
  });

  if (response.status === 401) {
    _cactusToken = null;
    throw new Error('Authentication failed. Please update your cactusToken.');
  } else if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`HTTP ${response.status}: ${errorText}`);
  }

  const responseBody = await response.json();
  
  if (responseBody.error) {
    throw new Error(`API Error: ${responseBody.error.message}`);
  }
  
  const predictions = responseBody.predictions;
  if (!predictions || predictions.length === 0) {
    throw new Error('No predictions in response');
  }
  
  const embeddings = predictions[0].embeddings;
  const values = embeddings.values;
  
  return values;
}

export async function getVertexAICompletion(
  textPrompt: string,
  imageData?: string,
  imagePath?: string,
  mimeType?: string,
): Promise<string> {
  if (_cactusToken === null) {
    throw new Error('CactusToken not set. Please call CactusVLM.init with cactusToken parameter.');
  }

  const projectId = 'cactus-v1-452518';
  const location = 'global';
  const modelId = 'gemini-2.5-flash-lite-preview-06-17';
  
  const endpoint = `https://aiplatform.googleapis.com/v1/projects/${projectId}/locations/${location}/publishers/google/models/${modelId}:generateContent`;

  const headers = {
    'Authorization': `Bearer ${_cactusToken}`,
    'Content-Type': 'application/json',
  };

  const parts: any[] = [];
  
  if (imageData) {
    const detectedMimeType = mimeType || 'image/jpeg';
    parts.push({
      inlineData: {
        mimeType: detectedMimeType,
        data: imageData
      }
    });
  } else if (imagePath) {
    const detectedMimeType = mimeType || detectMimeType(imagePath);
    const RNFS = require('react-native-fs');
    const base64Data = await RNFS.readFile(imagePath, 'base64');
    parts.push({
      inlineData: {
        mimeType: detectedMimeType,
        data: base64Data
      }
    });
  }
  
  parts.push({ text: textPrompt });

  const requestBody = {
    contents: {
      role: 'user',
      parts: parts,
    }
  };

  const response = await fetch(endpoint, {
    method: 'POST',
    headers,
    body: JSON.stringify(requestBody),
  });

  if (response.status === 401) {
    _cactusToken = null;
    throw new Error('Authentication failed. Please update your cactusToken.');
  } else if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`HTTP ${response.status}: ${errorText}`);
  }

  const responseBody = await response.json();
  
  if (Array.isArray(responseBody)) {
    throw new Error('Unexpected response format: received array instead of object');
  }
  
  if (responseBody.error) {
    throw new Error(`API Error: ${responseBody.error.message}`);
  }
  
  const candidates = responseBody.candidates;
  if (!candidates || candidates.length === 0) {
    throw new Error('No candidates in response');
  }
  
  const content = candidates[0].content;
  const responseParts = content.parts;
  if (!responseParts || responseParts.length === 0) {
    throw new Error('No parts in response');
  }
  
  return responseParts[0].text || '';
}

export async function getTextCompletion(prompt: string): Promise<string> {
  return getVertexAICompletion(prompt);
}

export async function getVisionCompletion(prompt: string, imagePath: string): Promise<string> {
  return getVertexAICompletion(prompt, undefined, imagePath);
}

export async function getVisionCompletionFromData(prompt: string, imageData: string, mimeType?: string): Promise<string> {
  return getVertexAICompletion(prompt, imageData, undefined, mimeType);
}

function detectMimeType(filePath: string): string {
  const extension = filePath.toLowerCase().split('.').pop();
  switch (extension) {
    case 'jpg':
    case 'jpeg':
      return 'image/jpeg';
    case 'png':
      return 'image/png';
    case 'gif':
      return 'image/gif';
    case 'webp':
      return 'image/webp';
    default:
      return 'image/jpeg';
  }
} 